### ISSUE TITLE
Dispatching responses inside of middleware

### ISSUE THREAD

**Issue by Dartv** [hkwu/ghastly#1]
Hey, awesome library! I enjoy it so far. I really like declarative style compared to other similar libraries.

But today I stuck when trying to figure out how to handle errors inside of middleware.
Assuming that I have `ErrorResponse` class that looks like this:
```js
import { Response } from 'ghastly/command';
import { RichEmbed } from 'discord.js';

import { COLOR_ERROR } from '../../util/constants';

export default class ErrorResponse extends Response {
  constructor(error) {
    super(async ({ message }) => {
      const embed = new RichEmbed();
      embed
        .setColor(COLOR_ERROR)
        .setTitle('🚫 Something went wrong!')
        .setDescription(error);
      return message.channel.send({ embed });
    });
  }
}
```
I can easily return it inside of handler:
```js
const handler = async () => new ErrorResponse('Permission denied!');
```
or
```js
const handler = async ({ dispatch }) => {
  await dispatch(new ErrorResponse('Permission denied!'));
};
```
That works fine, but what if I want to handle errors inside of middleware? I've noticed that context passed to middleware doesn't have `dispatch` property and just returning response does nothing. While debugging I found out that I can use `context.client.dispatcher.dispatchResponse(context, response)` inside of middleware but it just doesn't feel right to write it every time.

So, how exactly do I handle errors inside of middleware?

**Comment by hkwu** [comment]
Yes, layers don't get a `dispatch()` function since it's actually created when each command is called (in order to have a reference to the context object the handler receives).

Don't think it's possible to dispatch using response types easily in *before* middlewares atm. *After* middlewares get a context which has the `response` property set, so you could change that property to modify a command's response. But returning a raw `Response` in *before* middlewares is going to cause problems because the dispatcher isn't getting the context it expects.

Some solutions/hacks I thought of:
- Dispatch responses manually.
- Do the `dispatchResponse()` thing (though that's really supposed to be a private method).
- Create the response context manually (in your middleware), like

```js
return {
  response: new ErrorResponse('blah'),
  [Symbol.for('ghastly.originalContext')]: myContext,
};
```

I haven't maintained this lib in a while and haven't kept up with d.js, surprised it still works lol. I might be able to add a check and dispatch based on the return type, but tbh I think the cleanest solution is to ask users to explicitly pass the context to the response. I may very well change it to work like that since it would clean up the context handling, but it is a breaking change so let me know what you think.

```js
export default class ErrorResponse extends Response {
  constructor({ message }, error) {
    super(async () => {
      const embed = new RichEmbed();
      embed
        .setColor(COLOR_ERROR)
        .setTitle('🚫 Something went wrong!')
        .setDescription(error);
      return message.channel.send({ embed });
    });
  }
}
```

**Comment by Dartv** [comment]
Yeah, I think passing the context is a right thing to do since it feels "magical" on how the context comes in the first place.
Another solution I can think of is passing a fake `dispatch` function to middleware that will call `dispatchResponse` and then reassigning it to a real one. I don't know if it's possible though since I haven't looked at the source code much and I think your solution is much cleaner.

**Comment by hkwu** [comment]
Implemented in v0.4.0.
